### PR TITLE
feat: implement EXTCODECOPY opcode

### LINKED ISSUE #42: Opcode: EXTCODECOPY

### PR DESCRIPTION
**Pull request by ricomateo**
Closes #42 
This PR implements code generation for the [EXTCODECOPY](https://www.evm.codes/#3c?fork=cancun) opcode.

### CODE REVIEW

File: src/codegen/operations.rs [hunk 1]
```diff
@@ -3496,3 +3497,93 @@ fn codegen_chaind<'c, 'r>(
...
+fn codegen_extcodecopy<'c, 'r>(
+    op_ctx: &mut OperationCtx<'c>,
+    region: &'r Region<'c>,
+) -> Result<(BlockRef<'c, 'r>, BlockRef<'c, 'r>), CodegenError> {
+    let start_block = region.append_block(Block::new(&[]));
+    let context = &op_ctx.mlir_context;
+    let location = Location::unknown(context);
+    let uint32 = IntegerType::new(context, 32);
+
+    // TODO: add dynamic gas cost
```
Comment (TomasArrachea) [inline src/codegen/operations.rs:3510]: We can remove this TODO since the cost is implemented. Instead, we can add one for the cold dynamic cost that is missing.

**Review by TomasArrachea — APPROVED** [review]
LGTM
@MegaRedHand

File: src/syscall.rs [hunk 2]
```diff
@@ -311,6 +311,38 @@ impl<'c> SyscallContext<'c> {
...
+        dest_offset: u32,
+    ) {
+        let address_hi_bytes = address.hi.to_be_bytes();
+        let address_lo_bytes = address.lo.to_be_bytes();
+        // build the 20-byte address from the hi and lo bytes of the U256 address
+        let address = [&address_hi_bytes[12..16], &address_lo_bytes[..]].concat();
+        let address: Address = Address::from_slice(&address);
+        let code = match self.db.code_by_address(address) {
+            Ok(bytecode) => bytecode,
+            Err(_) => return,
```
Comment (MegaRedHand) [inline src/syscall.rs:329]: We should copy zeros when the account doesn't exist.

File: src/syscall.rs [hunk 3]
```diff
@@ -311,6 +311,38 @@ impl<'c> SyscallContext<'c> {
...
+        &mut self,
+        address: &U256,
+        code_offset: u32,
+        size: u32,
+        dest_offset: u32,
+    ) {
+        let address_hi_bytes = address.hi.to_be_bytes();
+        let address_lo_bytes = address.lo.to_be_bytes();
+        // build the 20-byte address from the hi and lo bytes of the U256 address
+        let address = [&address_hi_bytes[12..16], &address_lo_bytes[..]].concat();
```
Comment (MegaRedHand) [inline src/syscall.rs:325]: We should first check if the first 12 bits are set. In that case, the address is invalid and we write zeros

File: tests/evm.rs [hunk 4]
```diff
@@ -806,3 +806,82 @@ fn chainid_stack_overflow() {
...
+
+    assert!(&result.is_success());
+
+    let result_data = result.return_data().unwrap();
+    let expected_result = program.to_bytecode();
+    assert_eq!(result_data, &expected_result);
+}
+
+#[test]
+fn extcodecopy_with_offset_out_of_bounds() {
```
Comment (MegaRedHand) [inline tests/evm.rs:851]: Please add a test with an invalid address (first 12 bytes being non-zero)

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: tests/evm.rs [hunk 5]
```diff
@@ -885,3 +885,95 @@ fn extcodecopy_with_offset_out_of_bounds() {
...
+        Operation::Push((1_u8, BigUint::from(address))),
+        Operation::ExtcodeCopy, // 43th byte
+        Operation::Push((1_u8, BigUint::from(32_u8))),
+        Operation::Push((1_u8, BigUint::from(0_u8))),
+        Operation::Return,
+    ]
+    .into();
+
+    let mut env = Env::default();
+    env.tx.gas_limit = 999_999;
```
Comment (MegaRedHand) [inline tests/evm.rs:919]: This isn't needed since we fixed the default gas limit.

File: tests/evm.rs [hunk 6]
```diff
@@ -885,3 +885,95 @@ fn extcodecopy_with_offset_out_of_bounds() {
...
+        Operation::Push((32_u8, BigUint::from_bytes_be(wrong_address))),
+        Operation::ExtcodeCopy,
+        Operation::Push((1_u8, BigUint::from(size))),
+        Operation::Push((1_u8, BigUint::from(dest_offset))),
+        Operation::Return,
+    ]
+    .into();
+
+    let mut env = Env::default();
+    env.tx.gas_limit = 999_999;
```
Comment (MegaRedHand) [inline tests/evm.rs:964]: This too

File: src/syscall.rs [hunk 7]
```diff
@@ -44,16 +44,21 @@ impl U256 {
         self.lo = u128::from_be_bytes(buffer[16..32].try_into().unwrap());
         self.hi = u128::from_be_bytes(buffer[0..16].try_into().unwrap());
     }
-    pub fn try_into_address(&self) -> Option<Address> {
+}
+
+impl TryFrom<U256> for Address {
```
Comment (MegaRedHand) [inline src/syscall.rs:49]: Let's change this to a reference (`&U256`). That way we don't have to copy...

File: src/syscall.rs [hunk 8]
```diff
@@ -333,7 +338,7 @@ impl<'c> SyscallContext<'c> {
         let size = size as usize;
         let code_offset = code_offset as usize;
         let dest_offset = dest_offset as usize;
-        let Some(address) = address_value.try_into_address() else {
+        let Ok(address) = Address::try_from(*address_value) else {
```
Comment (MegaRedHand) [inline src/syscall.rs:341]: ...here

**Review by MegaRedHand — APPROVED** [review]
LGTM
